03_letters/pfannerj/main_test.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	mainout = &buf
+	main()
+	expected := `a:2
+b:1`
+	actual := buf.String()
+	if expected != actual {
+		t.Errorf(expected, actual, "Unexpected output in main()")
+	}
+}
+
+var lettersCases = map[string]struct {
+	input    string
+	expected map[rune]int
+}{
+	"Standard": {input: "aba", expected: map[rune]int{'a': 2, 'b': 1}},
+	"Same":     {input: "cccccccccccccccccccc", expected: map[rune]int{'c': 20}},
+	"Empty":    {input: "", expected: map[rune]int{}},
+	"Return":   {input: "ab\na\n", expected: map[rune]int{'\n': 2, 'a': 2, 'b': 1}},
+	"Slash":    {input: "ab/na/n", expected: map[rune]int{'/': 2, 'a': 2, 'b': 1, 'n': 2}},
+	"LongMixed": {input: "zabaaarzzzASYbcd@&^!*JHDJKHHGFGD^!*Dar",
+		expected: map[rune]int{'!': 2, '&': 1, '*': 2, '@': 1, 'A': 1, 'D': 3, 'F': 1, 'G': 2, 'H': 3, 'J': 2, 'K': 1, 'S': 1,
+			'Y': 1, '^': 2, 'a': 5, 'b': 2, 'c': 1, 'd': 1, 'r': 2, 'z': 4}},
+	"Punctuation": {input: ".:,!;'?.,:!'",
+		expected: map[rune]int{'!': 2, '\'': 2, ',': 2, '.': 2, ':': 2, ';': 1, '?': 1}},
+	"Raw": {input: `a:2b:1`, expected: map[rune]int{'1': 1, '2': 1, ':': 2, 'a': 1, 'b': 1}},
+	"RawReturn": {input: `a:2
+	b:1`, expected: map[rune]int{'	': 1, '\n': 1, '1': 1, '2': 1, ':': 2, 'a': 1, 'b': 1}},
+	"Numeric":      {input: "11235", expected: map[rune]int{'1': 2, '2': 1, '3': 1, '5': 1}},
+	"EscapedQuote": {input: "a\":2\"b:\"1", expected: map[rune]int{'"': 3, '1': 1, '2': 1, ':': 2, 'a': 1, 'b': 1}},
+	"RawQuote":     {input: `a":2"b:"1`, expected: map[rune]int{'"': 3, '1': 1, '2': 1, ':': 2, 'a': 1, 'b': 1}},
+}
```

Comment:
I know Ryan has already said "Really thorough test cases", but I'm going to counter that and say I think there are too many test cases here. Most of them are not testing anything different to the other test cases. For instance, `EscapedQuote` and `RawQuote` are the same test case - the different quoting is a property of the compiler, not a property of your code, so it makes little sense to have both tests. Ideally you have the smallest set of test cases to test your code.

Your code make no distinction between all the different types of characters you have in your tests, so you do not need all the tests that make that distinction. Every test is a test that needs to be validated and maintained so having too many tests can be a burden.

What is missing though is some unicode characters, where the value of the rune does not fit into a byte. When you index a string, you get a byte. When you range over it, you get runes (multi-byte characters). So when dealing with carving a string up into runes, it is worthwhile to have tests containing multi-byte characters to ensure you have not incorrectly indexed a string to get a byte, rather than correctly extracting a rune.

I would like to see some multi-byte character runes in the test cases before approving - I've seen one lab do this incorrectly (although I think it was numeronyms). I'll leave it up to you as to whether you want to reduce the size of the test cases or just take it into consideration for future labs.

The gory details about strings in Go, if you're interested: https://blog.golang.org/strings

Reply:
OK, I was a bit surprised to see that I don't have any multi-byte tests because I really thought I had included some. 
 I can see now that I did include them for the sortLetters tests but not for the other tests..